kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerRegistration.java

```diff
@@ -14,30 +14,87 @@
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.processSignals;
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.topicFromSignal;
+
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Consumer;
 
 import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.jbpm.kie.services.impl.KModuleDeploymentUnit;
 import org.jbpm.services.api.DeploymentEvent;
 import org.jbpm.services.api.model.DeployedAsset;
 import org.jbpm.services.api.model.MessageDesc;
 import org.jbpm.services.api.model.ProcessDefinition;
 import org.jbpm.services.api.model.SignalDesc;
 import org.jbpm.services.api.model.SignalDescBase;
 
-import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.processMessages;
-import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.processSignals;
-import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.topicFromSignal;
+import com.google.common.base.Objects;
 
 class KafkaServerRegistration {
 
-    private Map<String, Map<SignalDesc, Collection<String>>> topic2Signal = new HashMap<>();
-    private Map<String, Map<MessageDesc, Collection<String>>> topic2Message = new HashMap<>();
+    private static class DeploymentId {
+
+        private final String groupId;
+        private final String artifactId;
+        private final String version;
+
+        public DeploymentId(String groupId, String artifactId, String version) {
+            this.groupId = groupId;
+            this.artifactId = artifactId;
+            this.version = version;
+        }
+
+        public DeploymentId(String id) {
+            String[] tokens = id.split(":");
+            if (tokens.length >= 3) {
+                this.groupId = tokens[0];
+                this.artifactId = tokens[1];
+                this.version = tokens[2];
+            } else {
+                this.groupId = null;
+                this.artifactId = id;
+                this.version = null;
+            }
+        }
+
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+            if (groupId != null) {
+                sb.append(groupId).append(':');
+            }
+            sb.append(artifactId);
+            if (version != null) {
+                sb.append(':').append(version);
+            }
+            return sb.toString();
+        }
+
+        @Override
+        public int hashCode() {
+            return artifactId.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other instanceof DeploymentId) {
+                DeploymentId deployment = (DeploymentId) other;
+                return Objects.equal(this.groupId, deployment.groupId) && Objects.equal(this.artifactId, deployment.artifactId);
```

Comment:
This means will only have one deployment per signal so if we have 2 or more different versions of the same deployment it will happen that only one of them will get the signal processed. keep in mind that a process in version 1 won't go to vesion 2 because they belong to different kie containers, so it does not seems to me like a good idea.

Reply:
Since currently we are just notifying deployments (to start process instances) not process instance, the purpose of this JIRA is to avoid starting two process for the "same" deployment. 
https://kie.zulipchat.com/#narrow/stream/232678-jbpm/topic/Container.20upgrade.20and.20Kafka.20message